test/integration/src/main/java/io/pravega/test/integration/utils/SecureSetupUtils.java

```diff
@@ -0,0 +1,249 @@
+    @Getter
+    private EventStreamClientFactory clientFactory = null;
+    private ControllerWrapper controllerWrapper = null;
+    private PravegaConnectionListener server = null;
+    @Getter
+    private TestingServer zkTestServer = null;
+
+    @Getter
+    @Setter
+    private boolean authEnabled = false;
+    @Getter
+    @Setter
+    private boolean tlsEnabled = false;
+
+    // Manage the state of the class.
+    private final AtomicBoolean started = new AtomicBoolean(false);
+
+    // The test Scope name.
+    @Getter
+    private final String scope = "scope";
+    private final int controllerRPCPort = TestUtils.getAvailableListenPort();
+    private final int controllerRESTPort = TestUtils.getAvailableListenPort();
+    @Getter
+    private final int servicePort = TestUtils.getAvailableListenPort();
+    private ClientConfig clientConfig = null;
+
+    public ClientConfig generateValidClientConfig() {
+        ClientConfig.ClientConfigBuilder clientConfigBuilder = ClientConfig.builder()
+                .controllerURI(URI.create((this.tlsEnabled ? "tls" : "tcp") + "://localhost:" + controllerRPCPort));
+        if (this.authEnabled) {
+            clientConfigBuilder.credentials(new DefaultCredentials(SecurityConfigDefaults.AUTH_ADMIN_PASSWORD,
+                    SecurityConfigDefaults.AUTH_ADMIN_USERNAME));
+        }
+        if (this.tlsEnabled) {
+            clientConfigBuilder.trustStore("../" + SecurityConfigDefaults.TLS_CA_CERT_PATH)
+                    .validateHostName(false);
+        }
+        return clientConfigBuilder.build();
+    }
+
+    /**
+     * Start all pravega related services required for the test deployment.
+     *
+     * @throws Exception on any errors.
+     */
+    public void startAllServices() throws Exception {
+        startAllServices(null);
+    }
+
+    /**
+     * Start all pravega related services required for the test deployment.
+     *
+     * @param numThreads the number of threads for the internal client threadpool.
+     * @throws Exception on any errors.
+     */
+    public void startAllServices(Integer numThreads) throws Exception {
+        if (!this.started.compareAndSet(false, true)) {
+            log.warn("Services already started, not attempting to start again");
+            return;
+        }
+        this.clientConfig = generateValidClientConfig();
+        this.executor = ExecutorServiceHelpers.newScheduledThreadPool(2, "Controller pool");
+        this.controller = new ControllerImpl(ControllerImplConfig.builder().clientConfig(clientConfig).build(),
+                executor);
+        this.clientFactory = new ClientFactoryImpl(scope, controller, clientConfig);
+
+        // Start zookeeper.
+        this.zkTestServer = new TestingServerStarter().start();
+        this.zkTestServer.start();
+
+        // Start Pravega Service.
+        ServiceBuilder serviceBuilder = ServiceBuilder.newInMemoryBuilder(ServiceBuilderConfig.getDefaultConfig());
```

Comment:
You never shut this down. You need to keep a pointer to it and close it in your `close` method.

Reply:
I've added this as a field and close it in the `close` method